### PR TITLE
Added django_scim as a Django SCIM implementation, enabling queries on SCIM Users

### PR DESCRIPTION
**Pull request by QuanMPhm**
Addressing #29. I have added django_scim v0.19 as a dependency of the project. This module implements a large part of the SCIM specification for Django, and I've use this module to enable filter queries on users, such as filtering by email.

Depending on how needed it is, I can use the module to enable the same query functionality for Projects as well.

### CODE REVIEW

File: test-requirements.txt [hunk 1]
```diff
@@ -1 +1,2 @@
 git+https://github.com/nerc-project/coldfront-plugin-cloud@main#egg=coldfront_plugin_cloud
+git+https://github.com/15five/django-scim2@33e7caa
```
Comment (knikolla) [inline test-requirements.txt:2]: This is not necessary. You have already defined `django-scim2` as a dependency in `requirements.txt`.

File: src/coldfront_plugin_api/urls.py [hunk 2]
```diff
@@ -22,6 +23,7 @@ def get_queryset(self):
 
         return queryset
 
+app_name = 'scim'
```
Comment (knikolla) [inline src/coldfront_plugin_api/urls.py:26]: Why is this necessary?

File: src/coldfront_plugin_api/utils.py [hunk 3]
```diff
@@ -48,3 +48,6 @@ def get_or_fetch_user(username):
         user = create_user(**found)
 
     return user
+
+def SCIM_is_authenticated_predicate(user: User):
```
Comment (knikolla) [inline src/coldfront_plugin_api/utils.py:52]: A better name for this function is `is_user_superuser(user: User) -> bool` as that describes what this function is doing in this specific case, and therefore when you change the setting you get ```'GET_IS_AUTHENTICATED_PREDICATE': 'coldfront_plugin_api.utils.is_user_superuser'``` which describes the configuration setting without having to dig into the code.

File: src/coldfront_plugin_api/scim_v2/filters.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+'''
+Defines the SCIM filters for coldfront users and groups
+'''
+
+from django_scim.filters import UserFilterQuery, GroupFilterQuery
+from django_scim import constants
+from django_scim.utils import get_user_model
+
+class ColdfrontUserFilterQuery(UserFilterQuery):
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/filters.py:9]: Can you briefly explain what this and `attr_map` does?

**Review by knikolla — CHANGES_REQUESTED** [review]
Quick pass at reviewing, I'll give it a deeper look tomorrow. Overall it looks pretty good! Good work!

Some other comments
- You have 2 file that don't end with a newline. [It is a styling guideline to end all text files in a newline](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline) and you can configure your IDE to do that automatically so you don't have to worry about it. 
- If the `scim_v2/users.py` file is no longer required you should remove it. But instead I recommend you put the contenst of `adapters.py` into `users.py`. I don't have a strong reason for suggesting this, save for the fact that it may get bloated once we start adding group functionality, and dividing it by group and user in separate files feels cleaner and more highly related.

File: src/coldfront_plugin_api/urls.py [hunk 5]
```diff
@@ -31,7 +33,12 @@ def get_queryset(self):
 urlpatterns += [
     path('scim/v2/Groups', groups.ListGroups.as_view()),
     path('scim/v2/Groups/<int:pk>', groups.GroupDetail.as_view()),
-    path('scim/v2/Users', users.ListUsers.as_view()),
-    path('scim/v2/Users/<str:username>', users.UserDetail.as_view())
+    re_path(r'^scim/v2$',
```
Comment (larsks) [inline src/coldfront_plugin_api/urls.py:36]: I am unclear why we have replaced the `path(...)` expressions here with `re_path(...)` expressions, which are generally harder to read. Couldn't this be expressed something like:

```
path('scim/v2', scim_views.SCIMView.as_view(implemented=False), name='root'),
path('scvim/v2/Users', scim_views.UsersView.as_view(), name='users'),
path('scvim/v2/Users/<str:username>', scim_views.UsersView.as_view(), name='users'),
```

File: src/coldfront_plugin_api/scim_v2/filters.py (hunk 4 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/scim_v2/filters.py:9]: `UserFilterQuery` is the class used by `django-scim` to translate a SCIM filter query (i.e `filter=uesrName eq foo`) into a raw SQL query. 

For the translation to work, the query class needs to know which SCIM property maps to which SQL table column. This mapping must be defined with a `attr_map` dict (i.e I've mapped `emails.value` to `email`, so every SCIM filter query asking for `emails.value` will query the `email` field in the SQL database).

The internals of `UserFilterQuery` is related to the `scim2-filter-parser` module, made by the same organization. I mostly juts followed their [example](https://github.com/15five/scim2-filter-parser?tab=readme-ov-file#use)

File: src/coldfront_plugin_api/scim_v2/adapters.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    def from_dict(self, d):
+
+        '''
+        Overide from SCIMUser, this function first check if user exists through one of 
+        the configured search providers, and if yes, will fetch additional information from there.
+
+        Otherwise, for setting the attributes manually from the request's payload
+        '''
+        username = d.get('userName')
+        found = utils.find_user(username) # Fetch user
```
Comment (larsks) [inline src/coldfront_plugin_api/scim_v2/adapters.py:48]: We're using `d.get('username')`, which means that `username` could be `None`. Should we be calling `utils.find_user` in that case?

File: src/coldfront_plugin_api/urls.py (hunk 5 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py:36]: The view `django_scim.views.UsersView` is[ implemented in a fancy way ](https://github.com/15five/django-scim2/blob/33e7caa0c996d1977cb70a0f7463ba31d5ee24b8/src/django_scim/views.py#L288)where its behavior changes on whether `/<uuid>` was added to the URL or not. it also [hardcodes](https://github.com/15five/django-scim2/blob/33e7caa0c996d1977cb70a0f7463ba31d5ee24b8/src/django_scim/views.py#L37) the url arguement to be called `uuid`.

Comment (larsks) [reply src/coldfront_plugin_api/urls.py:36]: Would it work using `<str:uuid>`, then, or does it somehow require `re_path`?

File: src/coldfront_plugin_api/scim_v2/filters.py (hunk 4 above)
Comment (knikolla) [reply src/coldfront_plugin_api/scim_v2/filters.py:9]: Awesome, thanks for the explanation.

Please provide that as a class docstring, including the link too would be helpful. I'm generally not very fond of module level dosctrings.

Also, please use `"""` instead of `'''` for docstrings and be mindful of various other Python styling guidelines.

File: src/coldfront_plugin_api/scim_v2/adapters.py (hunk 6 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/scim_v2/adapters.py]: I've override the `validate_dict` function, which will check if the username field exists

File: src/coldfront_plugin_api/urls.py (hunk 5 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/urls.py]: Turns out you're right. The URLs still works fine if I split them and use `path`

**Comment by QuanMPhm** [comment]
>     * If the `scim_v2/users.py` file is no longer required you should remove it. But instead I recommend you put the contenst of `adapters.py` into `users.py`. I don't have a strong reason for suggesting this, save for the fact that it may get bloated once we start adding group functionality, and dividing it by group and user in separate files feels cleaner and more highly related.

I've put everything from `adapters.py` into `adapter_user.py`.

**Review by knikolla — COMMENTED** [review]
Can you also add a test that exercises the filering by "email" functionality?

Otherwise seems pretty good. And I'm okay with letting the unrelated linter changes that your pre-commit did to the files this once.

**Comment by QuanMPhm** [comment]
@knikolla I've added the test to check the filter functionality in `test_users.py`

File: src/coldfront_plugin_api/scim_v2/filters.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+from django_scim.filters import UserFilterQuery, GroupFilterQuery
+from django_scim import constants
+from django_scim.utils import get_user_model
+
+
+class ColdfrontUserFilterQuery(UserFilterQuery):
+    """
+    Defines the SCIM filters for Coldfront users
+
+    This class subclasses from `UserFilterQuery`, the class used by  `django-scim` to translate a SCIM filter query (i.e filter=uesrName eq foo) into a raw SQL query.
```
Comment (knikolla) [inline src/coldfront_plugin_api/scim_v2/filters.py:10]: Please wrap these lines. Most developers have `word wrap` turned off in their IDE/code editor and viewing this docstring would require scrolling sideways for them. Black uses a maximum of 88 characters per line, that is fine and we can stick with that.

Also make sure that you've removed the unused imports from the files that you've changed.

**Review by knikolla — CHANGES_REQUESTED** [review]
One last comment and I'll push this through.

File: src/coldfront_plugin_api/scim_v2/filters.py (hunk 7 above)
Comment (QuanMPhm) [reply src/coldfront_plugin_api/scim_v2/filters.py]: I've wrapped the docstring and removed the unused imports.

**Review by knikolla — APPROVED** [review]
